### PR TITLE
Start using lsif-typescript for JS/TS indexing

### PR DESCRIPTION
**Pull request by olafurpg**
This PR replaces lsif-node with the new lsif-typescript indexer for auto-indexing and the Sourcegraph CI.

## Test plan

I've manually tested these changes locally just to check that cross-repo navigation works as expected. The inference tests have been updated to reflect the new changes.

### CODE REVIEW

**Review by efritz — APPROVED** [review]

File: client/web/src/components/fuzzyFinder/FuzzyFinder.tsx [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import { ApolloError, useQuery } from '@apollo/client'
+// Hello trigger CI please
```
Comment (efritz) [inline client/web/src/components/fuzzyFinder/FuzzyFinder.tsx:2]:
```suggestion
```

File: lib/codeintel/autoindex/inference/typescript_test.go [hunk 2]
```diff
@@ -279,19 +279,19 @@ func TestInferTypeScriptIndexJobsTscLernaConfig(t *testing.T) {
...
-						Image:    lsifTscImage,
+						Image:    lsifTypescriptImage,
 						Commands: []string{"yarn --ignore-engines"},
 					},
 					{
 						Root:     "foo",
-						Image:    lsifTscImage,
-						Commands: []string{"yarn --ignore-engines"},
+						Image:    lsifTypescriptImage,
+						Commands: []string{"npm install"},
```
Comment (olafurpg) [inline lib/codeintel/autoindex/inference/typescript_test.go:288]: This change was suspicious. The test case has no `foo/yarn.lock` file so I'm not sure why it asserted that we used `yarn` before.

File: lib/codeintel/autoindex/inference/typescript.go [hunk 3]
```diff
@@ -127,33 +116,3 @@ func checkLernaFile(gitclient GitClient, path string, pathMap pathMap) (isYarn b
 	}
 	return false
 }
-
-func checkCanDeriveNodeVersion(gitclient GitClient, path string, pathMap pathMap) bool {
```
Comment (olafurpg) [inline lib/codeintel/autoindex/inference/typescript.go:131]: This logic got removed because 1) it seems unnecessary to install the exact version of Node and 2) it was very slow to download Node on every auto-indexing job

**Comment by olafurpg** [comment]
You can navigate the codebase with the new lsif-typescript indexes https://sourcegraph.com/github.com/sourcegraph/sourcegraph@9f95b3b1bc663e2e850385c5d4e4462887499ad8/-/blob/client/shared/src/settings/settings.ts?L110:5&subtree=true#tab=references

CleanShot 2022-03-15 at 09 29 56@2x

CleanShot 2022-03-15 at 09 32 45@2x

The signature in hover messages has different formatting from lsif-node. We'll probably try to update lsif-typescript to mirror the formatting eventually. There are still cases where lsif-typescript is missing hovers (wherever goto definition is failing, usually because cross-repo, cross-TS-project, or auto-generated sources) and we fallback to an older lsif-node upload.

File: lib/codeintel/autoindex/inference/typescript.go (hunk 3 above)
Comment (Strum355) [reply lib/codeintel/autoindex/inference/typescript.go:131]: This feels like it would be a regression in reliability if we removed it. My understanding is that it was specifically done to reduce a relatively high number of node version related errors (npm/node ecosystem is pretty finicky when it comes to this stuff and builds often depend on weakly/non- denoted version and system requirements).
If we want to try without for a while we could, in which case it might be easier to revert if it was in a separate PR.

Comment (olafurpg) [reply lib/codeintel/autoindex/inference/typescript.go:131]: I can add the logic back if it's needed. I don't understand why it would be needed though, we don't execute anything from the project (no scripts, no tests, no main methods). We only execute lsif-typescript.

Comment (olafurpg) [reply lib/codeintel/autoindex/inference/typescript.go:131]: > might be easier to revert if it was in a separate PR.

Good point, I'll separate it.

Comment (Strum355) [reply lib/codeintel/autoindex/inference/typescript.go:131]: It was most likely for `npm install`, where all the unwritten requirements come to light lol

Comment (olafurpg) [reply lib/codeintel/autoindex/inference/typescript.go:131]: We can probably work around the slow downloads by caching most major Node versions

Comment (olafurpg) [reply lib/codeintel/autoindex/inference/typescript.go]: Done, reverted the Node changes

File: .github/workflows/lsif-typescript.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      - name: Install npm dependencies
+        run: yarn --ignore-engines --ignore-scripts
+      - name: Install src-cli
+        run: |
+          curl -L https://sourcegraph.com/.api/src-cli/src_linux_amd64 -o /usr/local/bin/src
+          chmod +x /usr/local/bin/src
+      - name: Install lsif-typescript
+        run: yarn global add @sourcegraph/lsif-typescript
+      - name: Run lsif-typescript
+        run: lsif-typescript index --yarnWorkspaces
```
Comment (Strum355) [inline .github/workflows/lsif-typescript.yml:20]: Im assuming this removes the need for the root matrix?

**Review by Strum355 — APPROVED** [review]

File: .github/workflows/lsif-typescript.yml (hunk 4 above)
Comment (olafurpg) [reply .github/workflows/lsif-typescript.yml:20]: Correct, lsif-typescript invokes `yarn workspaces info` to list all workspaces and then indexes each of those in a for loop. We still need to fix a bit how to deal with cross-workspace navigation because it currently emits symbols for autogenerated definition files that are not checked into git. The TS build tool emits source maps that we could use to de-alias symbols from the definition files.

File: .github/workflows/lsif-typescript.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+name: lsif-typescript
+on:
+  push:
```
Comment (mrnugget) [inline .github/workflows/lsif-typescript.yml:3]: Why this change?

The original had this:

```
on:
  push:
    paths:
      - '**.ts'
      - '**.js'
```

so it only runs when `.ts` or `.js` files change and not on a docs-only or Go-only change. (We should probably include `tsx`, `jsx`?)

File: .github/workflows/lsif-typescript.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install npm dependencies
+        run: yarn --ignore-engines --ignore-scripts
+      - name: Install src-cli
+        run: |
+          curl -L https://sourcegraph.com/.api/src-cli/src_linux_amd64 -o /usr/local/bin/src
+          chmod +x /usr/local/bin/src
+      - name: Install lsif-typescript
+        run: yarn global add @sourcegraph/lsif-typescript
```
Comment (mrnugget) [inline .github/workflows/lsif-typescript.yml:18]: `yarn` is already pre installed?

Also: should we maybe dogfood the Docker image? The previous GitHub workflow used our Docker image with all the dependencies in it.

Comment (olafurpg) [reply .github/workflows/lsif-typescript.yml:18]: We can dogfood the docker container elsewhere. I prefer to keep the raw steps because they aren't that many and the base GH runners start faster in my experience.
